lualib/lua_scanners/avast_rest.lua

```diff
@@ -0,0 +1,354 @@
+        ['X-Correlation-Id'] = string.sub(task:get_uid(), 1,6)
+      }
+
+      request_data.url = make_url(addr, filename)
+    else
+
+      if rule.scan_mime_parts then
+
+        local _, _, mime_attr = maybe_part:get_type_full()
+        local det_ext = maybe_part:get_detected_ext()
+
+        if mime_attr.name then
+          local ext,_ = common.gen_extension(mime_attr.name)
+          if ext then
+            filename = string.format('file.%s', ext)
+          end
+        elseif det_ext then
+          filename = string.format('file.%s', det_ext)
+        else
+          filename = "file.txt"
+        end
+      else
+        filename = "file.eml"
+      end
+      request_data.url = make_url(addr, filename)
+      request_data.headers = {
+        ['Content-Type'] = 'application/octet-stream',
+        ['X-Correlation-Id'] = string.sub(task:get_uid(), 1,6)
+      }
+      request_data.body = content
+    end
+
+    lua_util.debugm(rule.name, task, '%s: request: url: %s, headers: %s, size: %s',
+      rule.log_prefix, request_data.url, request_data.headers, #request_data.body)
+
+    local function avast_rest_callback(http_err, code, body, headers)
+
+      local function requery()
+        -- set current upstream to fail because an error occurred
+        upstream:fail()
+
+        -- retry with another upstream until retransmits exceeds
+        if retransmits > 0 then
+
+          retransmits = retransmits - 1
+
+          lua_util.debugm(rule.name, task,
+              '%s: Request Error: %s - retries left: %s',
+              rule.log_prefix, http_err, retransmits)
+
+          -- Select a different upstream!
+          upstream = rule.upstreams:get_upstream_round_robin()
+          addr = upstream:get_addr()
+          request_data.url = make_url(addr, filename)
+
+          lua_util.debugm(rule.name, task, '%s: retry IP: %s:%s',
+              rule.log_prefix, addr, addr:get_port())
+
+          http.request(request_data)
+        else
+          rspamd_logger.errx(task, '%s: failed to scan, maximum retransmits '..
+              'exceed', rule.log_prefix)
+          task:insert_result(rule.symbol_fail, 0.0, 'failed to scan and '..
+              'retransmits exceed')
+        end
+      end
+
+      if http_err then
+        requery()
+      else
+        -- Parse the response
+        if upstream then upstream:ok() end
+
+        local cached = {}
+
+        if code ~= 200 then
+          rspamd_logger.errx(task, 'invalid HTTP code: %s, body: %s, headers: %s', code, body, headers)
+          common.yield_result(task, rule, string.format('Bad HTTP code: %s', code), 1.0, 'fail', maybe_part)
+          return
+        end
+        local data = string.gsub(tostring(body), '[\r\n%s]$', '')
+        lua_util.debugm(rule.name, task, '%s: got reply data: "%s"',
+            rule.log_prefix, data)
+
+        local ucl_parser = ucl.parser()
+        local ok, ucl_err = ucl_parser:parse_string(tostring(body))
```

Comment:
Same here, double `tostring` is double copy and interning. Please remove it for performance and sanity by replacing `parse_string` with `parse_text`.